upload-api/stores/agent/stream.js

```diff
@@ -0,0 +1,154 @@
+import * as UTF8 from 'uint8arrays/from-string'
+import * as API from '../../types.js'
+import * as Store from './store.js'
+
+export { API }
+
+export const defaults = {
+  workflow: { type: 'workflow' },
+  receipt: { type: 'receipt' },
+
+  // https://docs.aws.amazon.com/streams/latest/dev/key-concepts.html
+  // A partition key is used to group data by shard within a stream.
+  // It is required, and now we are starting with one shard. We need to study best partition key
+  partitionKey: 'key',
+}
+
+/**
+ * @typedef {import('@aws-sdk/client-kinesis').KinesisClientConfig} Address
+ * @typedef {Kinesis} Channel
+ *
+ * @typedef {API.Variant<{
+ *   channel: Channel
+ *   address: Address
+ *   disable: {}
+ * }>} Connection
+ *
+ * @typedef {object} Options
+ * @property {Connection} connection
+ * @property {string} name
+ * @property {string} [partitionKey]
+ * @property {{type:string}} [workflow]
+ * @property {{type:string}} [receipt]
+ *
+ * @typedef {object} Stream
+ * @property {Channel|null} channel
+ * @property {string} name
+ * @property {string} partitionKey
+ * @property {{type:string}} workflow
+ * @property {{type:string}} receipt
+ */
+
+/**
+ * @param {Options} options
+ * @returns {Stream}
+ */
+export const open = ({ connection, name, partitionKey, ...settings }) => ({
+  ...settings,
+  channel: connection.address
+    ? new Kinesis(connection.address)
+    : connection.channel ?? null,
+  name,
+  workflow: { ...defaults.workflow, ...settings.workflow },
+  receipt: { ...defaults.receipt, ...settings.receipt },
+  partitionKey: partitionKey ?? defaults.partitionKey,
+})
+
+/**
+ *
+ * @param {object} connection
+ * @param {Store.Store} connection.store
+ * @param {Stream} connection.stream
+ * @param {API.ParsedAgentMessage} message
+ * @returns {Promise<API.Result<API.Unit, Error>>}
+ */
+export const write = async (connection, message) => {
+  const { stream } = connection
+
+  try {
+    if (stream.channel) {
+      await stream.channel.putRecords({
+        Records: await assert(message, connection),
+        StreamName: connection.stream.name,
+      })
+    }
+    return { ok: {} }
+  } catch (cause) {
+    return { error: /** @type {Error} */ (cause) }
+  }
+}
+
+/**
+ * Iterates over all invocations and receipts and yields corresponding kinesis
+ * records.
+ *
+ * @param {API.ParsedAgentMessage} message
+ * @param {object} connection
+ * @param {Store.Store} connection.store
+ * @param {Stream} connection.stream
+ */
+export const assert = async (message, { stream, store }) => {
+  const records = []
+  for (const member of message.index) {
+    if (member.invocation) {
+      const { task, invocation, message } = member.invocation
+      records.push({
+        Data: UTF8.fromString(
+          JSON.stringify({
+            // This is bad naming but not worth a breaking change
+            carCid: message.toString(),
+            task: task.toString(),
```

Comment:
@vasco-santos I have realized only when updating tests a few minutes ago that this was called `invocationCid` and maybe something depends on this. My intention was to prepare things for a change where task and invocation cid aren't the same, which is why I put `task` link here and removed `invoactionCid` which appeared obsolete given that `value.cid` was the same.

I'm not sure what this was used for and whether it meant to refer to task or invocation, please let me know and we can change this as needed.

Reply:
I think invocationCID is used on several places like Billing that I am not familiar with. This also implies changes on the kinesis consumers then, and I would like to not tie changes in this PR so deep into other things. Can we keep this with the same and change those as follow up?